File: apps/desktop/src/marks.ts
/**
 * What a reader says about the game's wardrobe, as against what the game says about it.
 *
 * Everything else the transmog view draws is read out of the installed game, so it is the same
 * for everybody on that build. A mark is the one thing on that screen that is theirs: a star,
 * and tags they invented — "wishlist", "faction: horde", "what the alt wears". The storage is
 * `marks.rs` and `0016_transmog_marks.sql`; this is what a mark *means* once it is on screen,
 * and it is where the browsers ask their two questions of one.
 *
 * **A label and a property are one thing**, which is `TransmogTag.value` being nullable and
 * nothing more. A key with a value is a property, a key without one is a label, and both are
 * added by the same form, drawn as the same chip and filtered by the same picker. What that
 * buys is one feature where the issue asked for two; what it costs is remembering that `null`
 * means two different things in the two shapes here — on a [`TransmogTag`] it is "this tag has
 * no value", and on a [`TagChoice`] it is "any value, or none". They are both said out loud
 * where they are declared, because a reader meeting the second one first would guess wrong.
 *
 * The keys collate without case — see the migration — so every comparison of one key against
 * another in this file has to go through [`sameKey`] rather than `===`. A reader who typed
 * "Faction" once and "faction" later has one tag, and a filter that split them in two would
 * show them half of what they marked.
 */

import type { Facet } from "./terms";
import type { MarkSubjectKind, TransmogMark, TransmogMarksPayload, TransmogTag } from "./types";

/** A subject nobody has said anything about, which is nearly all of them. */
export const UNMARKED: TransmogMark = { kind: "set", id: 0, favourite: false, tags: [] };

/**
 * Every mark, indexed by the subject it is against.
 *
 * The payload is a list because that is what the two tables read as; a browser wants a lookup
 * per row it draws, and a page of a hundred looks scanning a list of four hundred marks is
 * forty thousand comparisons per keystroke of the search box.
 */
export interface MarkIndex {
  /** What this reader has said about one subject, or nothing at all. */
  of: (kind: MarkSubjectKind, id: number) => TransmogMark | undefined;
  /** How many subjects carry anything, which is what the browser says it is filtering over. */
  count: number;
}

/** One key of the index: the kind and the id together, because the two countings overlap. */
const at = (kind: MarkSubjectKind, id: number): string => `${kind}:${id}`;

export function indexMarks(payload: TransmogMarksPayload | null): MarkIndex {
  const bySubject = new Map<string, TransmogMark>();
  for (const mark of payload?.marks ?? []) bySubject.set(at(mark.kind, mark.id), mark);
  return {
    of: (kind, id) => bySubject.get(at(kind, id)),
    count: bySubject.size,
  };
}

/** Whether two keys are the same tag, which is `COLLATE NOCASE` in the migration. */
export function sameKey(left: string, right: string): boolean {
  return left.toLowerCase() === right.toLowerCase();
}

/** How a tag reads on a chip: the key alone where it is a label, and `key: value` otherwise. */
export function tagLabel(tag: TransmogTag): string {
  return tag.value === null ? tag.key : `${tag.key}: ${tag.value}`;
}

/**
 * One thing the tag picker above a browser can be set to.
 *
 * **A `value` of `null` here means "any value, or none"** — the choice a reader makes when they
 * want everything under a key, however they annotated it. That is deliberately not what `null`
 * means on the tag itself, where it means the tag has no value; a choice of `{ key, value:
 * null }` matches a bare label and a property alike, which is what "faction" as opposed to
 * "faction: horde" is asking for.
 */
export interface TagChoice {
  key: string;
  value: string | null;
  /** What the picker prints. */
  label: string;
  /** What the `<option>` carries, and what the view holds as its state. See [`tokenOf`]. */
  token: string;
}

/**
 * A choice as one string, so that a `<select>` and a piece of React state can carry it.
 *
 * The tab is the separator because a key and a value can hold anything a reader types except
 * whitespace that is not a plain space — `marks::clean_key` splits on whitespace and rejoins
 * with single spaces, so neither half can ever contain one. Anything more clever, a JSON blob
 * or an index into the list, either escapes badly or goes stale the moment an unrelated tag is
 * added somewhere else on the screen.
 */
export function tokenOf(key: string, value: string | null): string {
  return value === null ? key : `${key}\t${value}`;
}

/** The choice a token names, or nothing — which is the picker sitting on "any tag". */
export function choiceOf(token: string): { key: string; value: string | null } | null {
  if (!token) return null;
  const tab = token.indexOf("\t");
  if (tab < 0) return { key: token, value: null };
  return { key: token.slice(0, tab), value: token.slice(tab + 1) };
}

/**
 * Every tag in use against one kind of subject, as the picker offers them.
 *
 * A key that carries values earns a row of its own *and* a row per value: "faction" is how a
 * reader asks for both halves of a pair at once and "faction: horde" is how they ask for one.
 * A key only ever used as a label earns the one row, because "wishlist" and "wishlist, any
 * value" are the same question and offering both is offering a choice that does nothing.
 *
 * Sorted by key and then by value, both without case, so the list is the same list whatever
 * order the tags were invented in. The key printed is the first spelling the payload holds,
 * which is the one the backend stored last — a reader's own most recent correction.
 */
export function tagChoices(index: MarkIndex, kind: MarkSubjectKind, ids: number[]): TagChoice[] {
  const keys = new Map<string, { key: string; values: Map<string, string> }>();
  for (const id of ids) {
    for (const tag of index.of(kind, id)?.tags ?? []) {
      const folded = tag.key.toLowerCase();
      const held = keys.get(folded) ?? { key: tag.key, values: new Map<string, string>() };
      if (tag.value !== null) held.values.set(tag.value.toLowerCase(), tag.value);
      keys.set(folded, held);
    }
  }
  const choices: TagChoice[] = [];
  for (const folded of [...keys.keys()].sort()) {
    const held = keys.get(folded)!;
    choices.push({
      key: held.key,
      value: null,
      label: held.key,
      token: tokenOf(held.key, null),
    });
    for (const value of [...held.values.keys()].sort().map((one) => held.values.get(one)!)) {
      choices.push({
        key: held.key,
        value,
        label: `${held.key}: ${value}`,
        token: tokenOf(held.key, value),
      });
    }
  }
  return choices;
}

/**
 * What a browser is narrowed to beyond what the game says: the star, and one tag.
 *
 * One tag rather than several, because two tags at once is a query language and this is a
 * dropdown. A reader wanting "wishlist and not yet collected" has the free-text search, which
 * reads the tags too — see [`markWords`].
 */
export interface MarkFilter {
  /** Whether only the starred survive. */
  favourite: boolean;
  /** The token of the chosen tag, or `""` for any. See [`tokenOf`]. */
  tag: string;
}

/** A filter narrowing nothing, which is what both browsers open on. */
export const NO_MARK_FILTER: MarkFilter = { favourite: false, tag: "" };

/** Whether the filter would leave the list alone, which is what a count worth drawing is. */
export function marksNarrow(filter: MarkFilter): boolean {
  return filter.favourite || filter.tag !== "";
}

/**
 * Whether one subject's marks survive the filter.
 *
 * A subject nobody has marked survives an idle filter and nothing else, which is why the
 * unmarked case is `undefined` rather than an empty mark: "not starred" and "never touched"
 * filter identically, and only one of them needs a row in the database to say so.
 */
export function survivesMarks(mark: TransmogMark | undefined, filter: MarkFilter): boolean {
  if (filter.favourite && !mark?.favourite) return false;
  const wanted = choiceOf(filter.tag);
  if (!wanted) return true;
  return (mark?.tags ?? []).some((tag) => sameKey(tag.key, wanted.key)
    && (wanted.value === null || tag.value?.toLowerCase() === wanted.value.toLowerCase()));
}

/**
 * What a subject's marks add to what the free-text search reads, lowercased.
 *
 * So that typing "horde" finds what somebody filed under it without going near the picker,
 * which is the same argument the set search already makes for the expansion and the class: a
 * reader types the word they are thinking of, and the box is expected to know about it.
 */
export function markWords(mark: TransmogMark | undefined): string {
  if (!mark) return "";
  const tags = mark.tags.flatMap((tag) => (tag.value === null ? [tag.key] : [tag.key, tag.value]));
  // The word rather than a symbol, because a star is not typeable and "favourite" is what
  // somebody looking for their starred pieces would actually reach for.
  return [...(mark.favourite ? ["favourite"] : []), ...tags].join(" ").toLowerCase();
}

/**
 * And what they add to the terms it reads — one per tag, key and value as they were typed.
 *
 * [`markWords`]'s other half, and the reason a tag was a key and a value in the first place: a
 * label becomes a facet with no value, which is exactly what `wishlist:` asks for and what
 * `wishlist:soon` does not answer. Nothing about the star is here — "starred" is a checkbox above
 * the list rather than a word under a key, and the box beside it already reads [`markWords`].
 */
export function markFacets(mark: TransmogMark | undefined): Facet[] {
  return (mark?.tags ?? []).map((tag) => ({ key: tag.key, value: tag.value ?? "" }));
}

File: apps/desktop/src-tauri/src/marks.rs
//! What a person says about the game's wardrobe: a star, and tags of their own invention.
//!
//! The transmog view reads the installed game, so everything on it is true of every install
//! with that build. A mark is the one thing on that screen that is the reader's — "I want this
//! one", "this is the horde half of the pair", "this is what the alt wears" — and it is stored
//! in Chronie's own database because the game has nowhere to put it and no reason to.
//!
//! This module is the rules and nothing else: what a subject may be, what a key and a value
//! are allowed to look like, and what a stored mark is as a shape. The SQL is `collector.rs`,
//! beside every other statement in the app, and the window over it is `marks.ts`.
//!
//! **A label and a property are one thing here.** A key with a value is a property and a key
//! without one is a label, which is exactly the difference between `Some` and `None` — so
//! there is one editor, one table and one filter rather than two of each. What that costs is
//! one rule, [`clean_value`]: a value that cleans away to nothing becomes a label rather than
//! an empty string, because a tag whose value is `""` would filter like a property and read
//! like a label and be neither.

use serde::{Deserialize, Serialize};
use specta::Type;

/// A set the game ships, numbered by `TransmogSet.id`.
pub const SET: &str = "set";
/// A look, numbered by `ItemAppearance.id` — the game's own unit of collection.
pub const APPEARANCE: &str = "appearance";
/// A set the reader saved off the character themselves, numbered by this database.
///
/// The one subject of the three that Chronie issues the id for, and the reason marking one had
/// to be the same feature rather than a second one beside it: a set of somebody's own is a set,
/// and "star it, tag it, filter the browser by what you said" is what they already know how to
/// do to the ones Blizzard shipped. See `customsets::KIND`, which is this string from the other
/// end, and `0017_custom_sets.sql`, which widened the two tables to hold it.
pub const CUSTOM: &str = "custom";

/// How long a key may be. Long enough for "expansion" or "who wears it", short enough that a
/// chip stays a chip; a reader wanting a sentence about a look wants the value, not the key.
pub const KEY_LIMIT: usize = 48;

/// And how long a value may be. Four times the key, because "the one from the Timewalking
/// vendor in Tanaris" is a thing somebody will reasonably want to write.
pub const VALUE_LIMIT: usize = 160;

/// One thing somebody said about a look or a set.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct Tag {
    pub key: String,
    /// The value, or `None` where the key is the whole of what was said — which is a label.
    pub value: Option<String>,
}

/// Everything somebody has said about one subject, as the window reads it.
///
/// A mark exists only where there is something to say: a subject nobody has starred and nobody
/// has tagged has no row anywhere and no entry in a payload, which is what keeps this list the
/// length of what a person did rather than the length of the game's wardrobe.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct Mark {
    /// [`SET`], [`APPEARANCE`] or [`CUSTOM`].
    pub kind: String,
    /// The id of the thing: the game's for the first two, and this database's own for a set the
    /// reader saved. Which is why nothing here has a foreign key — two of the three subjects
    /// live in the game's files rather than in any table.
    pub id: i64,
    pub favourite: bool,
    pub tags: Vec<Tag>,
}

/// Every mark in the database, which is the whole of what the window is given.
///
/// All of them at once, and deliberately: this is what one person has said with their own
/// hands, so it is hundreds of rows rather than the fifty-five thousand looks they were said
/// about. Asking per set or per page would be four hundred round trips to save a payload that
/// is smaller than one set's icons.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct MarksPayload {
    pub marks: Vec<Mark>,
}

/// Which of the three kinds of subject this is, refusing anything that is none of them.
///
/// The check is here as well as in the table's own `CHECK` because the window sends this
/// across the bridge as a string, and "the database rejected it" is not a sentence anybody
/// should have to read to find out they meant `appearance`.
pub fn subject_kind(raw: &str) -> Result<&'static str, String> {
    match raw {
        SET => Ok(SET),
        APPEARANCE => Ok(APPEARANCE),
        CUSTOM => Ok(CUSTOM),
        other => Err(format!(
            "A mark belongs to a set, an appearance or a set of your own, not to a '{other}'."
        )),
    }
}

/// A subject id anything could have issued, which is any positive number.
///
/// Zero is what every hop of the transmog chain reads as when the game encrypts it — an
/// appearance belonging to content Blizzard has not shipped arrives as an id of nothing at
/// all — and a star against "whatever the game would not tell us about" is a star nobody can
/// ever find again.
pub fn subject_id(id: i64) -> Result<i64, String> {
    if id > 0 {
        Ok(id)
    } else {
        Err("The game says nothing about that one, so there is nothing to mark.".into())
    }
}

/// The key as it will be stored: what was typed, with the whitespace made ordinary.
///
/// Trimmed and with runs of space collapsed, because a key is compared against other keys and
/// "off  hand" and "off hand" being two tags is a distinction nobody typed on purpose. Control
/// characters go for the reason they go everywhere else in this app — a stored string that can
/// move a cursor is a string that draws wrong in every reader downstream.
///
/// The case is *kept*, and the uniqueness is case-insensitive: that is `COLLATE NOCASE` in the
/// migration. Somebody who types "Faction" gets "Faction" back, and typing "faction" later
/// edits the tag they already have rather than making a second one.
pub fn clean_key(raw: &str) -> Result<String, String> {
    let key = collapse(raw);
    if key.is_empty() {
        return Err("A tag needs a name.".into());
    }
    if key.chars().count() > KEY_LIMIT {
        return Err(format!(
            "A tag's name has to fit in {KEY_LIMIT} characters."
        ));
    }
    Ok(key)
}

/// The value as it will be stored, or `None` where what was typed says nothing.
///
/// Cleaned the same way the key is, and then the one rule that makes a label: a value that is
/// empty, or that was only ever spaces, is stored as no value rather than as `''`. That is
/// what makes "label" and "property" the same feature — see the module's own note — and it is
/// also what clearing a value means, which is the same act as never having typed one.
pub fn clean_value(raw: Option<&str>) -> Result<Option<String>, String> {
    let value = collapse(raw.unwrap_or_default());
    if value.is_empty() {
        return Ok(None);
    }
    if value.chars().count() > VALUE_LIMIT {
        return Err(format!("A tag's value has to fit in {VALUE_LIMIT} characters."));
    }
    Ok(Some(value))
}

/// Whitespace and control characters made ordinary: one space between words, none at the ends.
fn collapse(raw: &str) -> String {
    raw.split_whitespace()
        .map(|word| {
            word.chars()
                .filter(|character| !character.is_control())
                .collect::<String>()
        })
        .filter(|word| !word.is_empty())
        .collect::<Vec<_>>()
        .join(" ")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn recognises_the_three_kinds_of_subject_and_nothing_else() {
        assert_eq!(subject_kind("set"), Ok(SET));
        assert_eq!(subject_kind("appearance"), Ok(APPEARANCE));
        assert_eq!(subject_kind("custom"), Ok(CUSTOM));
        assert!(subject_kind("item").is_err());
        assert!(subject_kind("").is_err());
        // Case is not a spelling of the same word here: this crosses the bridge as one of two
        // literals the window writes, not as something anybody types.
        assert!(subject_kind("Set").is_err());
    }

    #[test]
    fn refuses_an_id_the_game_never_issued() {
        assert_eq!(subject_id(1), Ok(1));
        assert_eq!(subject_id(712245), Ok(712245));
        // Which is what an appearance the game encrypts arrives as.
        assert!(subject_id(0).is_err());
        assert!(subject_id(-3).is_err());
    }

    #[test]
    fn tidies_a_key_without_changing_what_was_typed() {
        assert_eq!(clean_key("  faction  "), Ok("faction".into()));
        assert_eq!(clean_key("off   hand"), Ok("off hand".into()));
        // The case survives; the migration is what makes two spellings of it one tag.
        assert_eq!(clean_key("Faction"), Ok("Faction".into()));
        assert_eq!(clean_key("who\twears\nit"), Ok("who wears it".into()));
        assert_eq!(clean_key("nor\u{7}mal"), Ok("normal".into()));
    }

    #[test]
    fn refuses_a_key_that_is_nothing_or_endless() {
        assert!(clean_key("").is_err());
        assert!(clean_key("   ").is_err());
        assert!(clean_key("\u{7}").is_err());
        assert!(clean_key(&"a".repeat(KEY_LIMIT)).is_ok());
        assert!(clean_key(&"a".repeat(KEY_LIMIT + 1)).is_err());
    }

    #[test]
    fn a_value_that_says_nothing_is_a_label() {
        assert_eq!(clean_value(None), Ok(None));
        assert_eq!(clean_value(Some("")), Ok(None));
        assert_eq!(clean_value(Some("   ")), Ok(None));
        assert_eq!(clean_value(Some("  horde ")), Ok(Some("horde".into())));
        assert_eq!(
            clean_value(Some("the one  from Tanaris")),
            Ok(Some("the one from Tanaris".into()))
        );
    }

    #[test]
    fn refuses_a_value_longer_than_a_chip_can_hold() {
        assert!(clean_value(Some(&"a".repeat(VALUE_LIMIT))).is_ok());
        assert!(clean_value(Some(&"a".repeat(VALUE_LIMIT + 1))).is_err());
    }

    /// A character outside the basic plane is one character, not four — the limits are about
    /// what a reader typed rather than about how many bytes it took to say it.
    #[test]
    fn counts_characters_rather_than_bytes() {
        assert!(clean_key(&"é".repeat(KEY_LIMIT)).is_ok());
        assert!(clean_value(Some(&"🐉".repeat(VALUE_LIMIT))).is_ok());
    }
}
